### PR TITLE
add calculix

### PR DESCRIPTION
**Pull request by looooo**

### CODE REVIEW

**Comment by looooo** [comment]
@mingwandroid What is the best way to add some headers for the mingw-toolchain? Somehow all my attempts to fix this are not working.
The error is:
```
gcc -Wall -O3 -fopenmp -I %PREFIX%\Library/include/spooles/ -DARCH="Linux" -DSPOOLES -DARPACK -DMATRIXSTORAGE -DUSE_MT=1 -c arpack.c
In file included from arpack.c:26:0:
spooles.h:26:18: fatal error: misc.h: No such file or directory
```

but misc.h is in %PREFIX%\Library/include/spooles/

File: recipes/calculix/Makefile_MT_linux [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+OCCXC = $(SCCXC:.c=.o)
+OCCXMAIN = $(SCCXMAIN:.c=.o)
+
+DIR=../../../SPOOLES.2.2
+
+LIBS = \
+       $(PREFIX)/lib/spoolesMT.a \
+       $(PREFIX)/lib/spooles.a \
+       $(PREFIX)/lib/libarpack.a \
+       $(PREFIX)/lib/liblapack.a \
```
Comment (isuruf) [inline recipes/calculix/Makefile_MT_linux]: Why use static libraries?

File: recipes/calculix/Makefile_MT_linux [hunk 2]
```diff
@@ -0,0 +1,35 @@
+
+CFLAGS = -Wall -O3 -fopenmp -I $(PREFIX)/include/spooles -DARCH="Linux" -DSPOOLES -DARPACK -DMATRIXSTORAGE -DUSE_MT=1 
+FFLAGS = -Wall -O3 -fopenmp 
+
+CC=gcc
+FC=gfortran
```
Comment (isuruf) [inline recipes/calculix/Makefile_MT_linux]: Don't hardcode FC and CC

File: recipes/calculix/Makefile_MT_linux [hunk 3]
```diff
@@ -0,0 +1,35 @@
+
+CFLAGS = -Wall -O3 -fopenmp -I $(PREFIX)/include/spooles -DARCH="Linux" -DSPOOLES -DARPACK -DMATRIXSTORAGE -DUSE_MT=1
```
Comment (isuruf) [inline recipes/calculix/Makefile_MT_linux]: Don't override these. Use the values from the environment and append

File: recipes/calculix/Makefile_MT_linux [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+       $(PREFIX)/lib/spooles.a \
+       $(PREFIX)/lib/libarpack.a \
+       $(PREFIX)/lib/liblapack.a \
+       -lpthread -lm
+
+ccx_2.14_MT: $(OCCXMAIN) ccx_2.14_MT.a  $(LIBS)
+	./date.pl; $(CC) $(CFLAGS) -c ccx_2.14.c; $(FC) -fopenmp -Wall -O3 -o $@ $(OCCXMAIN) ccx_2.14_MT.a $(LIBS)
+
+ccx_2.14_MT.a: $(OCCXF) $(OCCXC)
+	ar vr $@ $?
```
Comment (isuruf) [inline recipes/calculix/Makefile_MT_linux]: Don't hardcode `ar`

**Comment by mingwandroid** [comment]
@looooo Make is an MSYS2 package so it doesn't understand `\`. Please see [this](https://github.com/conda/conda-recipes/tree/master/msys2/m2w64-muparser) as an example (bld.bat in particular).

**Comment by mingwandroid** [comment]
I like the following pattern at the top of makefiles:

AR ?= ar

This allows a default if the env var is not set.

File: recipes/calculix/Makefile_MT_windows [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+LIBS = \
+       $(LIBRARY_PREFIX)/lib/spoolesMT.a \
+       $(LIBRARY_PREFIX)/lib/spooles.a \
+       $(LIBRARY_PREFIX)/mingw-w64/lib/libarpack.a \
+       $(BUILD_PREFIX)/_build_env/Library/mingw-w64/x86_64-w64-mingw32/lib/libpthread.a \
+       $(BUILD_PREFIX)/_build_env/Library/mingw-w64/x86_64-w64-mingw32/lib/libm.a
+
+ccx_2.14_MT: $(OCCXMAIN) ccx_2.14_MT.a  $(LIBS)
+	./date.pl; $(CC) $(CFLAGS) -c ccx_2.14.c; $(FC) -static-libgcc -static-libgfortran  -fopenmp -Wall -O3 -o $@ $(OCCXMAIN) ccx_2.14_MT.a $(LIBS)
```
Comment (isuruf) [inline recipes/calculix/Makefile_MT_windows]: Why static libgfortran and libgcc?

File: recipes/calculix/Makefile_MT_windows [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+OCCXF = $(SCCXF:.f=.o)
+OCCXC = $(SCCXC:.c=.o)
+OCCXMAIN = $(SCCXMAIN:.c=.o)
+
+DIR=..\..\..\SPOOLES.2.2
+
+LIBS = \
+       $(LIBRARY_PREFIX)/lib/spoolesMT.a \
+       $(LIBRARY_PREFIX)/lib/spooles.a \
+       $(LIBRARY_PREFIX)/mingw-w64/lib/libarpack.a \
```
Comment (isuruf) [inline recipes/calculix/Makefile_MT_windows]: Why static here?

Comment (looooo) [reply recipes/calculix/Makefile_MT_windows]: There are not many references for this library, but it was suggested here: https://mechanicalhacks.wordpress.com/2011/03/29/building-calculix-to-run-native-on-64-bit-windows/
https://mechanicalhacks.wordpress.com/2011/03/29/building-calculix-to-run-native-on-64-bit-windows/

File: recipes/calculix/bld.bat [hunk 7]
```diff
@@ -0,0 +1,14 @@
+cd ccx*/src
+
+rm Makefile_MT
+CP %RECIPE_DIR%\Makefile_MT Makefile_MT
+CP %BUILD_PREFIX%\Library\mingw-w64\bin\mingw32-make %BUILD_PREFIX%\Library\mingw-w64\bin\make
```
Comment (looooo) [inline recipes/calculix/bld.bat]: is there a better way to get a make-command? I guess the source-code should use $(MAKE) instead of make.

**Comment by looooo** [comment]
errors:
- win: 
```
f77 -Wall -O3 -fopenmp -c absolute_relative.f
process_begin: CreateProcess(NULL, f77 -Wall -O3 -fopenmp -c absolute_relative.f, ...) failed.
make (e=2): The system cannot find the file specified.```
- osx:
```
conda_build.exceptions.DependencyNeedsBuildingError: Unsatisfiable dependencies for platform osx-64: {'*[track_features=blas_openblas]', 'arpack=3.6.3'}
```

**Comment by looooo** [comment]
locally the build now also fails for linux:
```
/opt/conda/conda-bld/calculix_1543677923461/_build_env/bin/x86_64-conda_cos6-linux-gnu-ar: absolute_relative.o: No such file or directory
```

Previously the build already worked for linux.

File: recipes/calculix/Makefile_MT_windows (hunk 6 above)
Comment (looooo) [reply recipes/calculix/Makefile_MT_windows]: @isuruf I guess arpack is only available as a static library for windows. And it's better to have consistent run-deps for the other os.

**Comment by looooo** [comment]
@mingwandroid I have now successfully build calculix on windows. But installing it downgrades the full environment. The only dependency of this package is m2w64-gcc-libs. So I guess the downgrading is caused by this dependency.

So my questions are:
1. Is it possible to make this package not depend on any library? I guess it should be possible as I link only to static libraries.
2. if this isn't possible, update the m2w64-gcc-libs stack, so no downgrading occurs.

**Comment by mingwandroid** [comment]
The msys2 packages shouldn't cause any downgrades! They're used extensively in defaults.

File: recipes/calculix/bld.bat [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+echo "------------------------"
+echo %LIBRARY_PREFIX%
+echo "------------------------"
+
+make -f Makefile_MT ^
+    SPOOLES_INCLUDE_DIR="%LIBRARY_PREFIX%/mingw-w64/include/spooles" ^
+    LIB_DIR="%LIBRARY_PREFIX%/mingw-w64/lib" ^
+    FORTRAN_COMPILER="gfortran"
+
+cp ccx_*_MT "%LIBRARY_PREFIX%/bin/ccx"
```
Comment (looooo) [inline recipes/calculix/bld.bat]: should be `cp ccx_*_MT "%LIBRARY_PREFIX%/bin/ccx.exe"`

**Comment by looooo** [comment]
> 
> 
> The msys2 packages shouldn't cause any downgrades! They're used extensively in defaults.

Maybe it's related to some other stuff currently going on in conda-forge. Let's ignore it for now ; )

File: recipes/calculix/Makefile_MT_windows (hunk 6 above)
Comment (isuruf) [reply recipes/calculix/Makefile_MT_windows]: Please send a PR to arpack to build it as a shared library.

**Comment by looooo** [comment]
thanks @isuruf for the help. Do you have any idea what the problem for linux and osx is? Was there anything changed for the dependency-solver recently?

File: recipes/calculix/bld.bat [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+
+REM this line translates the windows-paths to paths understandable for the mingw env
+REM -m, --mixed           like --windows, but with regular slashes (C:/WINNT)
+FOR /F "delims=\" %%i IN ('cygpath.exe -m "%LIBRARY_PREFIX%"') DO set "LIBRARY_PREFIX=%%i"
+
+make -f Makefile_MT ^
+    SPOOLES_INCLUDE_DIR="%LIBRARY_PREFIX%/mingw-w64/include/spooles" ^
+    LIB_DIR="%LIBRARY_PREFIX%/mingw-w64/lib" ^
+    FC="gfortran"
```
Comment (looooo) [inline recipes/calculix/bld.bat]: @isuruf it doesn't work without setting the fortran compiler explicitly. If I do not set this f77 is used which is not available. Is this a problem with the mingw environment?

File: recipes/calculix/Makefile_MT [hunk 10]
```diff
@@ -0,0 +1,30 @@
+AR ?= ar
+FC ?= gfortran
```
Comment (isuruf) [inline recipes/calculix/Makefile_MT]: @looooo, FC is set here.

Comment (looooo) [reply recipes/calculix/Makefile_MT]: @isuruf No without setting it explicitly in the bld.bat it didn't work and f77 was used.

**Comment by looooo** [comment]
@conda-forge/staged-recipes I think this is ready for feedstock.

File: recipes/calculix/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    commands:
+        - test -f ${PREFIX}/bin/ccx  # [unix]
+        - test -f %LIBRARY_PREFIX%\bin\ccx.exe  # [win]
+
+
+about:
+    home: http://www.dhondt.de/
+    license: GPL 2
+    license_family: GPL
+    # license_file: COPYING
```
Comment (scopatz) [inline recipes/calculix/meta.yaml]: GPL code requires the license.  You may need to put the license in the recipe.  See https://conda-forge.org/docs/meta.html#packaging-the-license-manually

Comment (looooo) [reply recipes/calculix/meta.yaml]: thanks, done.

Comment (looooo) [reply recipes/calculix/meta.yaml]: also contacted the author to apply patches and add license file for next version.

**Pull request merged**